### DISCUSSION TITLE
討論有關於「pkexec」和「PATH」

### DISCUSSION THREAD

**Discussion by samwhelp** [LingmoOS/Community#25] [Bugs Report/缺陷提交] [answered]
# 討論有關於「pkexec」和「PATH」

## 緣起

我也有遇到在另一個「[Issue](https://github.com/LingmoOS/lingmo-core/issues/11#issue-2553057704)」描述的狀況。

三不五時就會跳出如下圖的對話框。

![pkexec](https://github.com/user-attachments/assets/22219c53-445d-43e7-88a1-a17fb5358234)

> 所以稍微研究了一下，紀錄如下

## 環境

* [Lingmo OS 2.0 Hydrogen](https://github.com/orgs/LingmoOS/discussions/20)
* lingmo-core    2.0.1

## /usr/share/polkit-1/actions/com.lingmo.brightness.pkexec.policy

執行下面指令

``` sh
cat /usr/share/polkit-1/actions/com.lingmo.brightness.pkexec.policy
```

顯示

``` xml
<?xml version="1.0" encoding="UTF-8"?> <!DOCTYPE policyconfig PUBLIC
 "-//freedesktop//DTD PolicyKit Policy Configuration 1.0//EN"
 "http://www.freedesktop.org/standards/PolicyKit/1/policyconfig.dtd">
<policyconfig>
  <action id="com.lingmo.brightness.pkexec">
    <message>Authentication is required to Change Brightness</message>
    <icon_name>battery</icon_name>
    <defaults>
      <allow_any>no</allow_any>
      <allow_inactive>no</allow_inactive>
      <allow_active>yes</allow_active>
    </defaults>
    <annotate key="org.freedesktop.policykit.exec.path">/usr/bin/lingmo-screen-brightness</annotate>
  </action>
</policyconfig>
```

> 要注意的是「`<annotate key="org.freedesktop.policykit.exec.path">/usr/bin/lingmo-screen-brightness</annotate>`」那一行，

> 是設定「`/usr/bin/lingmo-screen-brightness`」。

## 測試「pkexec」

執行下面指令，就會出現上面那個圖的對話框。

``` sh
pkexec lingmo-screen-brightness --set 100
```

> 對話框的提示訊息如下

```
Authentication is needed to run `/bin/lingmo-screen-brightness --set 100` as the super user
```

> 要注意的是「`/bin/lingmo-screen-brightness`」。

於是改執行下面指令，就會如預期正常運作

``` sh
pkexec /usr/bin/lingmo-screen-brightness --set 100
```

``` sh
pkexec /usr/bin/lingmo-screen-brightness --set 90
```

> 也就是要確切的指定「`/usr/bin/lingmo-screen-brightness`」，就會符合上面「`<annotate key="org.freedesktop.policykit.exec.path">/usr/bin/lingmo-screen-brightness</annotate>`」那一行的設定。

## lingmo-core/settings-daemon/brightness/brightnessmanager.cpp

我有稍微找了一下實作的程式碼，找到如下

程式碼片段： [lingmo-core/settings-daemon/brightness/brightnessmanager.cpp](https://github.com/LingmoOS/lingmo-core/blob/main/settings-daemon/brightness/brightnessmanager.cpp#L80)

``` cpp
    process.start("pkexec", QStringList() << "lingmo-screen-brightness" << "--set" << QString::number(value));
```

> 也就是符合上面的測試，目前的實作，會執行「`pkexec lingmo-screen-brightness --set 100`」

> 然後在加上目前的「PATH」設定，導致就會變成「`pkexec /bin/lingmo-screen-brightness --set 100`」，於是就會出現一開始那張圖顯示的對話框。

> 接下來說明關於「Lingmo」，目前在「PATH」的設定

## 測試「PATH」

執行下面指令

``` sh
echo $PATH
```

顯示

```
/home/sam/.local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/local/bin
```

> 上面「PATH」的設定：「`/bin`」優先於「`/usr/bin`」。

執行下面指令

``` sh
file /bin
```

顯示

```
/bin: symbolic link to usr/bin
```

> 上面表示「/bin」軟連結到「/usr/bin」。

> 也就是「/bin」和「/usr/bin」裡面存在的指令是一樣的。

執行

``` sh
which lingmo-screen-brightness
```

顯示

```
/bin/lingmo-screen-brightness
```

執行

``` sh
which ls
```

顯示

```
/bin/ls
```

以上的測試，就可以了解到為何會發生上面提到的狀況

執行下面指令

``` sh
pkexec lingmo-screen-brightness --set 100
```

就會出先一個對話框，提示訊息如下

```
Authentication is needed to run `/bin/lingmo-screen-brightness --set 100` as the super user
```

> 這也會導致在「Status Bar」執行「螢幕亮度調整」時，無法正常運作，甚至會讓系統產生「卡頓」的情形。

## 關於「Debian」原本的「PATH」設定

找了一下「Debian」原本的「PATH」設定

可以找到「/etc/profile」開頭其中有一段「程式碼片段」如下

``` sh
if [ "$(id -u)" -eq 0 ]; then
  PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
else
  PATH="/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games"
fi
export PATH
```

在「Debian」執行下面指令

``` sh
echo $PATH
```

顯示

```
/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
```

> 從上面可以了解到，在「Debian」，「/usr/bin」的採用順序，是優先於「/bin」。

## 關於「Lingmo」目前的「PATH」設定

在「Lingmo」執行下面指令

``` sh
echo $PATH
```

顯示

```
/home/sam/.local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/local/bin
```

> 從上面可以了解到，在「Lingmo」，「/bin」的採用順序，是優先於「/usr/bin」。

> 程式碼片段：[/etc/profile](https://github.com/LingmoOS/lingmo-base-common/blob/main/share/profile#L3-L31)

``` sh
# Append "$1" to $PATH when not already in.
# This function API is accessible to scripts in /etc/profile.d
append_path () {
    case ":$PATH:" in
        *:"$1":*)
            ;;
        *)
            PATH="${PATH:+$PATH:}$1"
    esac
}

# Append our default paths
append_path '/usr/local/sbin'
append_path '/usr/local/bin'
append_path '/usr/bin'

# Force PATH to be environment
export PATH

# Load profiles from /etc/profile.d
if test -d /etc/profile.d/; then
        for profile in /etc/profile.d/*.sh; do
                test -r "$profile" && . "$profile"
        done
        unset profile
fi

# Unload our profile API functions
unset -f append_path
```

> 我做了點測試，修改「/etc/profile」程式碼片段如下

``` sh
echo "start: ${PATH}" > /tmp/log.txt

# Append our default paths
append_path '/usr/local/sbin'
append_path '/usr/local/bin'
append_path '/usr/bin'

echo "end: ${PATH}" >> /tmp/log.txt
```

重新開機後，執行下面指令，

``` sh
cat /tmp/log.txt
```

觀看「`/tmp/log.txt`」內容如下

```
start: /bin:/usr/bin
end: /bin:/usr/bin:/usr/local/sbin:/usr/local/bin
```

> 表示在「append_path」之前，一開始「PATH」，就已經是「`/bin:/usr/bin`」。

> 若是要調整的話，在一開頭，應該先將「`PATH`」清空，再開始「append_path」，類似如下

``` sh
echo "start: ${PATH}" > /tmp/log.txt

# Init PATH
PATH=""

echo "init: ${PATH}" >> /tmp/log.txt

# Append our default paths
append_path '/usr/local/bin'
append_path '/usr/local/sbin'
append_path '/usr/bin'
append_path '/bin'

echo "end: ${PATH}" >> /tmp/log.txt
```

重新開機後，執行下面指令，

``` sh
cat /tmp/log.txt
```

觀看「`/tmp/log.txt`」內容如下

```
start: /bin:/usr/bin
init:
end: /usr/local/bin:/usr/local/sbin:/usr/bin:/bin
```

另外也修改了一個「prepend」的版本，程式碼片段如下。

``` sh
# Prepend "$1" to $PATH when not already in.
# This function API is accessible to scripts in /etc/profile.d
prepend_path () {
    case ":$PATH:" in
        *:"$1":*)
            ;;
        *)
            PATH="$1${PATH:+:$PATH}"
    esac
}

# Init PATH
PATH=""

# Prepend our default paths
prepend_path '/bin'
prepend_path '/usr/bin'
prepend_path '/usr/local/sbin'
prepend_path '/usr/local/bin'


# Force PATH to be environment
export PATH



# Unload our profile API functions
unset -f prepend_path
```

> 修改「/etc/profile」範例

| 修改範例 | 修改後的「/etc/profile」|
| ------- | --------------------- |
| [append-path](https://github.com/samwhelp/lingmo-adjustment/tree/main/sample/demo/profile-config/append-path) | [/etc/profile](https://github.com/samwhelp/lingmo-adjustment/blob/main/sample/demo/profile-config/append-path/asset/overlay/etc/profile#L3-L35) |
| [prepend-path](https://github.com/samwhelp/lingmo-adjustment/tree/main/sample/demo/profile-config/prepend-path) | [/etc/profile](https://github.com/samwhelp/lingmo-adjustment/blob/main/sample/demo/profile-config/prepend-path/asset/overlay/etc/profile#L3-L53) |

## 建議

> 統合以上提到的，建議需要調整的地方，有以下兩點。

1. [程式實作](https://github.com/LingmoOS/lingmo-core/blob/main/settings-daemon/brightness/brightnessmanager.cpp#L80)時，執行「`pkexec /usr/bin/lingmo-screen-brightness --set 100`」，要指定確切路徑「`/usr/bin/lingmo-screen-brightness`」。
2. [調整「PATH」的採用順序](https://github.com/LingmoOS/lingmo-base-common/blob/main/share/profile#L3-L31)，「/usr/bin」優先於「/bin」。

## 後記

* [回覆#30: Screen brightness problem ](https://github.com/orgs/LingmoOS/discussions/30)

**Comment by chun-awa** [comment, marked as answer]
最开始我还不信他跟profile有关，查了一大堆文件（甚至对比原cutefish的brightness模块，polkit rule，etc.）最后发现确实是PATH的问题（

以下为正常的Debian 13的PATH：
```console
root@f6cfc8cc7b01:/# echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@f6cfc8cc7b01:/#
```
（Arch Linux的PATH里甚至没有`/bin`）

事实证明，**`/bin`本该在后面**

**Comment by elysia-best** [comment]
Fixed in https://github.com/LingmoOS/lingmo-core/pull/18

**Comment by samwhelp** [comment]
> 補充紀錄一個跟「PATH」這個議題相關的發現

在「[上面](https://github.com/orgs/LingmoOS/discussions/25#discussion-7293264)」我有提到，我做了點測試

發現到，在進入「/etc/profile」的流程前，

一開始的「`PATH`」就已經被設定「`/bin:/usr/bin`」，

我剛剛在「[探索sddm](https://github.com/LingmoOS/lingmo-sddm-theme/issues/4)」時，

執行「[sddm --example-config](https://samwhelp.github.io/note-about-sddm/read/command/sddm.html#example-config)」後，發現到其中一個片段如下


``` ini
[Users]
# Default $PATH for logged in users
DefaultPath=/bin:/usr/bin

```

> 上面的片段，也可以執行「`sddm --example-config | grep '\[Users\]' -A 2`」來找到。


> 因此我聯想到，這也就能解釋，為啥一開始的「`PATH`」就已經被設定「`/bin:/usr/bin`」，是被「sddm」設定的。


我們可以來做個小實驗，編輯「`/etc/sddm.conf`」，修改如下

``` ini
[Users]
# Default $PATH for logged in users
DefaultPath=/bin:/usr/bin:/opt/bin
```

> 我們多加了一個「`/opt/bin`」在後面

然後我們先登出，重新登入，開啟「Terminal」，執行「`echo $PATH`」，就可以看到多加入了「`/opt/bin`」。


以上補充，報告完畢

**Reply by elysia-best** [reply]
Thanks~
